### PR TITLE
[ Application ] mnist with tensorflow

### PR DESCRIPTION
**Pull request by jijoongmoon**
This PR includes:
  . MNIST Tensorflow Example to compare with nntrainer.

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

**Comment by jijoongmoon** [comment]
image

File: Applications/mnist/Tensorflow/Training_Keras.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+    model_Hist = History_LAW()
+
+    history = model.fit(datagen(InVec, InLabel, batch_size), epochs=num_epoch, steps_per_epoch=len(InVec)//batch_size, validation_data=datagen(ValVec, ValLabel, batch_size), validation_steps=len(ValVec)//batch_size, callbacks=[model_Hist])
+    
+    count =0
+    
+    if not os.path.exists('./nntrainer_tfmodel'):
+        os.mkdir('./nntrainer_tfmodel')
+        model.save('./nntrainer_tfmodel/nntrainer_keras.h5')
```
Comment (zhoonit) [inline Applications/mnist/Tensorflow/Training_Keras.py:147]: Model is not saved when `os.path.exists('nntrainer_tfmodel')`
Maybe this?
```suggestion
    model.save('./nntrainer_tfmodel/nntrainer_keras.h5')
```

File: Applications/mnist/Tensorflow/Training_Keras.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+import random
+import struct
+import os
+import tensorflow as tf
+import numpy as np
+from tensorflow.keras.layers import Input, Dense, Conv2D, Flatten, AveragePooling2D
+from tensorflow.keras import models, layers, optimizers
+from tensorflow.keras.callbacks import Callback
+
```
Comment (kparichay) [inline Applications/mnist/Tensorflow/Training_Keras.py:27]: lets take deterministic code making sections from input_gen.py
or may be add them to a separate file and use it everywhere to be able to reproduce values from tensorflow.

File: Applications/mnist/Tensorflow/Training_Keras.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+            y_batch = y_data[i*batch_size: (i+1)*batch_size]
+            
+
+            yield x_batch, y_batch
+            
+##
+# @brief load input data from file
+# @return (InputVector, InputLabel, Validation Vector, ValidationLabel)
+def load_data():
+    tf.compat.v1.set_random_seed(777)
```
Comment (kparichay) [inline Applications/mnist/Tensorflow/Training_Keras.py:77]: this will go away once above comment is applied.

File: Applications/mnist/Tensorflow/Training_Keras.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    for i in range(TOTAL_LABEL_SIZE*TOTAL_VAL_DATA_SIZE):
+        for j in range(FEATURE_SIZE):
+            data_str = fin.read(4)
+            ValVector[i,j] = struct.unpack('f',data_str)[0]
+        for j in range(TOTAL_LABEL_SIZE):
+            data_str = fin.read(4)
+            ValLabel[i,j] = struct.unpack('f',data_str)[0]
+    fin.close()
+
+    fin=open('mnist_trainingSet.dat','rb')
```
Comment (kparichay) [inline Applications/mnist/Tensorflow/Training_Keras.py:98]: so in this case, input and validation are same?
if yes, please add a comment here on this. Otherwise this might look like a bug later.

File: Applications/mnist/Tensorflow/Training_Keras.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+    model.compile(optimizer = optimizers.Adam(lr=1e-4),
+                  loss=tf.keras.losses.CategoricalCrossentropy(),
+                  metrics=['accuracy'])
+
+    model.summary()
+
+    model_Hist = History_LAW()
+
+    history = model.fit(datagen(InVec, InLabel, batch_size), epochs=num_epoch, steps_per_epoch=len(InVec)//batch_size, validation_data=datagen(ValVec, ValLabel, batch_size), validation_steps=len(ValVec)//batch_size, callbacks=[model_Hist])
```
Comment (kparichay) [inline Applications/mnist/Tensorflow/Training_Keras.py:141]: either shuffle false or remove shuffle from datagen.

Comment (kparichay) [inline Applications/mnist/Tensorflow/Training_Keras.py:141]: The input data passed to model.fit is not normalized. However, the inputlayer in mnist.ini has normalization turned on.
Lets make it consistent when making comparison.

Comment (jijoongmoon) [reply Applications/mnist/Tensorflow/Training_Keras.py:97]: done. Thanks

Comment (jijoongmoon) [reply Applications/mnist/Tensorflow/Training_Keras.py:97]: done. Thanks

File: Applications/mnist/Tensorflow/Training_Keras.py (hunk 4 above)
Comment (jijoongmoon) [reply Applications/mnist/Tensorflow/Training_Keras.py]: done. Thanks

File: Applications/mnist/Tensorflow/Training_Keras.py (hunk 3 above)
Comment (jijoongmoon) [reply Applications/mnist/Tensorflow/Training_Keras.py]: done. Thanks

File: Applications/mnist/Tensorflow/Training_Keras.py (hunk 2 above)
Comment (jijoongmoon) [reply Applications/mnist/Tensorflow/Training_Keras.py:29]: done. Thanks

File: Applications/mnist/Tensorflow/Training_Keras.py (hunk 1 above)
Comment (jijoongmoon) [reply Applications/mnist/Tensorflow/Training_Keras.py:103]: done. Thanks

File: Applications/mnist/Tensorflow/Training_Keras.py (hunk 2 above)
Comment (kparichay) [reply Applications/mnist/Tensorflow/Training_Keras.py:29]: Add `os.environ['TF_CUDNN_DETERMINISTIC'] = '1'` before import tf

Add below below here.
```
# Fix the seeds across frameworks
SEED = ....
tf.compat.v1.reset_default_graph()
random.seed(SEED)
tf.compat.v1.set_random_seed(SEED)
np.random.seed(SEED)
```

Comment (jijoongmoon) [reply Applications/mnist/Tensorflow/Training_Keras.py:31]: done
